pkg/config/mesh/watcher.go

```diff
@@ -76,29 +81,28 @@ func NewWatcher(fileWatcher filewatcher.FileWatcher, filename string) (Watcher,
 			log.Warnf("failed to read mesh configuration, using default: %v", err)
 			return
 		}
+		w.handleMeshConfig(meshConfig)
+	})
+	return w, nil
+}
 
-		var handlers []func()
-
-		w.mutex.Lock()
-		if !reflect.DeepEqual(meshConfig, w.mesh) {
-			log.Infof("mesh configuration updated to: %s", spew.Sdump(meshConfig))
-			if !reflect.DeepEqual(meshConfig.ConfigSources, w.mesh.ConfigSources) {
-				log.Infof("mesh configuration sources have changed")
-				//TODO Need to re-create or reload initConfigController()
-			}
-
-			// Store the new mesh.
-			atomic.StorePointer((*unsafe.Pointer)(unsafe.Pointer(&w.mesh)), unsafe.Pointer(meshConfig))
-			handlers = append([]func(){}, w.handlers...)
-		}
-		w.mutex.Unlock()
-
-		// Notify the handlers of the change.
-		for _, h := range handlers {
-			h()
+// NewConfigMapWatcher creates a new Watcher for changes to the given ConfigMap.
+func NewConfigMapWatcher(client kube.Client, namespace, name, key string) Watcher {
+	defaultMesh := DefaultMeshConfig()
+	w := &watcher{mesh: &defaultMesh}
```

Comment:
I think we need to wait for it to sync somewhere. Otherwise at startup we may temporary have the wrong (default) mesh config which can cause serious impact

Reply:
Sounds good. From my understanding, there can be a short gap between the underlying cache sync and the first handler call, so I'll plan to do what the secret controller does by not marking an initial sync until the first queue item is processed